kimchi/src/circuits/polynomials/chacha.rs

```diff
@@ -371,63 +223,233 @@ pub fn constraint<F: FftField>(alpha0: usize) -> E<F> {
             alpha0,
             vec![
                 // booleanity of overflow bit
-                boolean(&x_plus_z_overflow_bit),
+                Self::boolean(&x_plus_z_overflow_bit),
                 // x' = x + z (mod 2^32)
-                combine_nybbles(x_plus_z_nybbles) - xprime.clone(),
+                Self::combine_nybbles(x_plus_z_nybbles) - xprime.clone(),
                 // Correctness of x+z nybbles
                 xprime + E::from(1 << 32) * x_plus_z_overflow_bit - (x + z),
                 // Correctness of y nybbles
-                combine_nybbles(y_nybbles) - y,
+                Self::combine_nybbles(y_nybbles) - y,
                 // y' = (y ^ x') <<< 4 * nybble_rotation
-                combine_nybbles(y_xor_xprime_rotated) - yprime,
+                Self::combine_nybbles(y_xor_xprime_rotated) - yprime,
             ],
         )
-    };
-
-    let chacha_final = {
-        let y_xor_xprime_nybbles = chunks_over_2_rows(1);
-        let low_bits = chunks_over_2_rows(5);
-        let yprime = w(0, Curr);
+    }
+}
 
-        let one_half = F::from(2u64).inverse().unwrap();
+/// a += b; d ^= a; d <<<= 16 (=4*4)
+pub fn constraint_chacha0<F: FftField>(alpha0: usize) -> E<F> {
+    let index = |g: GateType| E::cell(Column::Index(g), Curr);
+    index(GateType::ChaCha0) * Impl::line(alpha0, 4)
```

Comment:
Duplicated code here isn't ideal. It would be preferable to have
```rust
pub fn constraint_shift_by_nybbles<F: FftField>(alpha0: usize, gate_type: GateType) -> E<F> {
    let line_number =
        match gate_type {
            GateType::ChaCha0 => 4,
            GateType::ChaCha1 => 3,
            GateType::ChaCha2 => 2,
            _ => panic!("The gate type {} is not a nybble-shifting ChaCha gate!", gate_type)
        };
    let index = |g: GateType| E::cell(Column::Index(g), Curr);
    index(gate_type) * Impl::line(alpha0, line_number)
}
```

Reply:
The idea is to make this more explicit, so I think it's better to maintain a 3 lines of duplicated code